crypto/core_namemap.c

```diff
@@ -11,44 +11,133 @@
 #include <openssl/lhash.h>
 #include <openssl/safestack.h>
 
-/* The namemap entry */
+/*-
+ * The numname entry
+ * =================
+ *
+ * This entry associates a number with a set of names, i.e. what's needed to
+ * have a 1-to-many number-to-name database.
+ *
+ * +----------------+
+ * | names          | ----------+
+ * +----------------+           |
+ * | names_n        |           |
+ * +----------------+           |
+ * | names_sz       |           |
+ * +----------------+           |
+ * | body           | ---------------+
+ * +----------------+           |    |
+ * | body_n         |           |    |
+ * +----------------+           |    |
+ * | body_sz        |           |    |
+ * +----------------+           |    |
+ *                              |    |
+ *                              |    |
+ *                              |    |
+ * +----------------+           |    |
+ * | ptr1           | ---+   <--+    |
+ * +----------------+    |           |
+ * | ptr2           | -----+         |
+ * +----------------+    | |         |
+ * :  ...           :    : :         :
+ * +----------------+    | |         |
+ * | ptrn           | -------+       |
+ * +----------------+    | | |       |
+ * | NULL           |    | | |       |
+ * +----------------+    | | |       |
+ *                       | | |       |
+ *                       | | |       |
+ *                       | | |       |
+ * +----------------+    | | |       |
+ * | "name1"        | <--+ | |    <--+
+ * +----------------+      | |
+ * | "name2"        | <----+ |
+ * +----------------+        |
+ * :  ...           :        :
+ * +----------------+        |
+ * | "namen"        | <------+
+ * +----------------+
+ *
+ * These are then used in a stack indexed by number (for number to
+ * names mapping)
+ */
+
 typedef struct {
+    /* names and n_names are organized the same way as argv and argc */
+    const char **names;
```

Comment:
Would make more sense to use a stack here?

Perhaps also for the body but that's slightly further from the stack behaviour.